In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge
from sklearn.metrics import mean_squared_error

# Load dataset
file_path = "BostonHousing.csv"  # Update with your file path
df = pd.read_csv(file_path)

# Assuming 'Price' is the target variable
X = df.drop(columns=['medv'])  # Features
y = df['medv']  # Target

# Train-validation split (80% train, 20% validation)
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

### (a) Train Linear Regression
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Predictions
y_pred_linear = linear_model.predict(X_val)

# Compute MSE
mse_linear = mean_squared_error(y_val, y_pred_linear)
print(f"Linear Regression MSE: {mse_linear:.4f}")

# Print feature coefficients
print("\nLinear Regression Coefficients:")
for feature, coef in zip(X.columns, linear_model.coef_):
    print(f"{feature}: {coef:.4f}")

### (b) Train Lasso (L1 Regularization)
lasso_model = Lasso(alpha=0.1)  # Alpha is the regularization strength
lasso_model.fit(X_train, y_train)

# Predictions
y_pred_lasso = lasso_model.predict(X_val)

# Compute MSE
mse_lasso = mean_squared_error(y_val, y_pred_lasso)
print(f"\nLasso Regression MSE: {mse_lasso:.4f}")

# Print feature coefficients
print("\nLasso Regression Coefficients:")
for feature, coef in zip(X.columns, lasso_model.coef_):
    print(f"{feature}: {coef:.4f}")

### (c) Train Ridge (L2 Regularization)
ridge_model = Ridge(alpha=0.1)
ridge_model.fit(X_train, y_train)

# Predictions
y_pred_ridge = ridge_model.predict(X_val)

# Compute MSE
mse_ridge = mean_squared_error(y_val, y_pred_ridge)
print(f"\nRidge Regression MSE: {mse_ridge:.4f}")

# Print feature coefficients
print("\nRidge Regression Coefficients:")
for feature, coef in zip(X.columns, ridge_model.coef_):
    print(f"{feature}: {coef:.4f}")


Linear Regression MSE: 24.2911

Linear Regression Coefficients:
crim: -0.1131
zn: 0.0301
indus: 0.0404
chas: 2.7844
nox: -17.2026
rm: 4.4388
age: -0.0063
dis: -1.4479
rad: 0.2624
tax: -0.0106
ptratio: -0.9155
b: 0.0124
lstat: -0.5086

Lasso Regression MSE: 25.1556

Lasso Regression Coefficients:
crim: -0.1042
zn: 0.0349
indus: -0.0168
chas: 0.9200
nox: -0.0000
rm: 4.3117
age: -0.0151
dis: -1.1515
rad: 0.2392
tax: -0.0130
ptratio: -0.7322
b: 0.0131
lstat: -0.5647

Ridge Regression MSE: 24.3010

Ridge Regression Coefficients:
crim: -0.1124
zn: 0.0305
indus: 0.0349
chas: 2.7503
nox: -15.9245
rm: 4.4458
age: -0.0073
dis: -1.4296
rad: 0.2600
tax: -0.0108
ptratio: -0.9008
b: 0.0124
lstat: -0.5109


### Conclusion Based on Linear, Lasso, and Ridge Regression Results

####  Linear Regression (Baseline Model)

MSE = 24.2911 (Lowest among all models)

No regularization is applied, so it keeps all features with their original influence.

Some coefficients have large absolute values (e.g., nox: -17.2026, rm: 4.4388), indicating strong influence.

Susceptible to overfitting since it does not penalize complex models.

#### Lasso Regression (L1 Regularization)

MSE = 25.1556 (Slightly higher than Linear Regression)

L1 regularization removes some coefficients completely (e.g., nox becomes 0).

Helps with feature selection by shrinking some features to exactly zero.

Reduces model complexity but may cause underfitting when too many important features are removed.

### Ridge Regression (L2 Regularization)

MSE = 24.3010 (Very close to Linear Regression, slightly higher)

L2 regularization shrinks coefficients but does not eliminate them.

Helps prevent overfitting while retaining all features in the model.

Compared to Lasso, Ridge retains nox but with a smaller magnitude (-15.9245 instead of -17.2026).